### PR TITLE
Fixed Read More Button Display Logic to Avoid Unnecessary Display

### LINKED ISSUE #9329: "Read more" button should not be added if it takes up as much space as the hidden text
#### Evidence / Screenshot

image
image

#### Relevant URL(s)

E.g. https://openlibrary.org/books/OL8978501M/Robin_Hood

### Reproducing the bug

1. Go to https://openlibrary.org/books/OL8978501M/Robin_Hood

* Expected behavior: Description not collapsed, since "read more" makes the section actually longer :P
* Actual behavior: "Read more" shown over description

### Context

- Browser (Chrome, Safari, Firefox, etc): FF
- OS (Windows, Mac, etc): Win
- Logged in (Y/N): Y
- Environment (prod, dev, local): prod

#### Proposal & constraints
Perhaps we can update the `+ 1` here: https://github.com/internetarchive/openlibrary/blob/73b70f6b997c401caf2fe8078a6296c8f3167ee5/openlibrary/plugins/openlibrary/js/readmore.js#L58

To instead be the height of the `.read-more__toggle--more` button. We might need to capture that height in the constructor though, before the read more button is potentially hidden!

#### Related files
https://github.com/internetarchive/openlibrary/blob/73b70f6b997c401caf2fe8078a6296c8f3167ee5/openlibrary/plugins/openlibrary/js/readmore.js

#### Stakeholders
@RayBB 

**Note:** Before [making a new branch](https://github.com/internetarchive/openlibrary/wiki/Git-Cheat-Sheet#making-changes-and-creating-a-pull-request) _or_ updating an existing one, please [ensure your branch is up to date](https://github.com/internetarchive/openlibrary/wiki/Git-Cheat-Sheet#working-on-your-branch).

### PR DESCRIPTION
**Pull request by CodeMaverick2**
Closes #9329
## Description

This PR addresses an issue where the "Read more" button was displayed even when it made the section longer than the fully expanded content. This led to a confusing user experience, as the "Read more" button was not serving its intended purpose. The following changes have been made to resolve this issue:

## Changes

1. **Capture "Read More" Button Height**:
    - The height of the "Read more" button is now captured during initialization to ensure accurate calculations.

2. **Update Condition in `reset` Method**:
    - The logic that determines whether the "Read more" button should be displayed has been updated to include the height of the button. This ensures the button is only shown when it makes sense.

## Impact

- **Improved User Experience**: The "Read more" button is now only displayed when necessary, preventing unnecessary clutter and confusion.
- **Correct Layout Behavior**: Ensures the layout behaves as expected, with the "Read more" button enhancing usability rather than detracting from it.

## Testing
To test visit a link like https://testing.openlibrary.org/books/OL8978501M/Robin_Hood

## Screenshots
Screenshot from 2024-07-25 18-43-42

### CODE REVIEW

**Comment by CodeMaverick2** [comment]
@RayBB I have done the changes recommended by @cdrini

**Review by RayBB — APPROVED** [review]
@CodeMaverick2  I pushed up a small change so that the tests pass and shouldn't impact functionality. Otherwise this looks good.

Unfortunately, my deploys to testing seem to be failing and I'm not sure why. 
However, I've deployed it to gitpod and it's working for me.

Lets just get a final check from @cdrini

**Comment by CodeMaverick2** [comment]
@RayBB  Sure, @cdrini Can u check the problem

**Comment by CodeMaverick2** [comment]
@RayBB Are there any more changes that need to be done in this

**Comment by RayBB** [comment]
We are waiting for staff to review. They often do this on Monday. If not feel free to tag @cdrini again

**Comment by CodeMaverick2** [comment]
@cdrini What's the status here ?

File: openlibrary/plugins/openlibrary/js/readmore.js [hunk 1]
```diff
@@ -55,7 +55,7 @@ export class ReadMoreComponent {
         this.fullHeight = this.$content.scrollHeight;
         // Fudge factor to account for non-significant read/more
         // (e.g missing a bit of padding)
-        if (this.$content.scrollHeight <= (this.collapsedHeight + 1)) {
+        if (this.$content.scrollHeight <= (this.collapsedHeight + (this.$readMoreButton?.offsetHeight | 0) + 1)) {
```
Comment (cdrini) [inline openlibrary/plugins/openlibrary/js/readmore.js:58]: Nice this is working almost perfectly ! A few tweaks:

- We'll want to compute and _store_ the offsetHeight of the readMore button the first time reset is called. Otherwise when the read more button is hidden and `reset` is called, it will switch between existing and not existing. (You can observe this by trying to resize the window on testing. You'll see the read more button appear and disappear, since sometimes the height is there and sometimes it's zero). The first time reset is called, the read more is always visible, so you can likely do something like: `this.readMoreHeight ||= this.$readMoreButton.offsetHeight;` at the top of this method.

- Also we don't need the `?.` since it's always defined, and we don't need to default to zero since `offsetHeight` will always be a valid number.
- (also: Small typo, I think you mean `||` instead of `|` 😁 )

**Review by cdrini — CHANGES_REQUESTED** [review]
One small bug, otherwise looking great! 😊

File: openlibrary/plugins/openlibrary/js/readmore.js (hunk 1 above)
Comment (CodeMaverick2) [reply openlibrary/plugins/openlibrary/js/readmore.js]: Thank you for pointing that out—I’ve just fixed it!

**Comment by CodeMaverick2** [comment]
@RayBB  @cdrini The tests are failing like previously not able to figure out why

File: openlibrary/plugins/openlibrary/js/readmore.js (hunk 1 above)
Comment (cdrini) [reply openlibrary/plugins/openlibrary/js/readmore.js]: We'll need to implement that first bullet point as well!

**Comment by cdrini** [comment]
Nice! There's still the first point; we'll want to fetch the height value only once (see the first bullet point)

Oh the tests are failing cause it looks `readMoreButton` is `null` in the tests; You'll need to update the test so that it is not null :+1:
